core/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java

```diff
@@ -0,0 +1,156 @@
+            refreshListenersEstimatedSize++;
+            return;
+        }
+        /*
+         * No free slot so force a refresh and call the listener in this thread. Do so outside of the synchronized block so any other
+         * attempts to add a listener can continue.
+         */
+        forceRefresh.run();
+        listener.accept(true);
+    }
+
+    /**
+     * Start listening to an engine.
+     */
+    public void listenTo(Engine engine) {
+        engine.registerSearchRefreshListener(new RefreshListenerCallingRefreshListener(engine));
+    }
+
+    /**
+     * Listens to Lucene's {@linkplain ReferenceManager.RefreshListener} and fires off listeners added by
+     * {@linkplain IndexShard#addRefreshListener(Translog.Location, Consumer)}.
+     */
+    private class RefreshListenerCallingRefreshListener implements ReferenceManager.RefreshListener {
+        private final Engine engine;
+        private Translog.Location currentRefreshLocation;
+
+        public RefreshListenerCallingRefreshListener(Engine engine) {
+            this.engine = engine;
+        }
+
+        @Override
+        public void beforeRefresh() throws IOException {
+            currentRefreshLocation = engine.getTranslog().getLastWriteLocation();
+        }
+
+        @Override
+        public void afterRefresh(boolean didRefresh) throws IOException {
+            // This intentionally ignores didRefresh so a refresh call over the API can force rechecking the refreshListeners.
+            if (null == currentRefreshLocation) {
+                /*
+                 * The translog had an empty last write location at the start of the refresh so we can't alert anyone to anything. This
+                 * usually happens during recovery. The next refresh cycle out to pick up this refresh.
+                 */
+                return;
+            }
+            /*
+             * First set the lastRefreshedLocation so listeners that come in locations before that will just execute inline without messing
+             * around with refreshListeners at all.
+             */
+            lastRefreshedLocation = currentRefreshLocation;
+            /*
+             * Now pop all listeners off the front of refreshListeners that are ready to be called. The listeners won't always be in order
+             * but they should be pretty close because you don't listen to times super far in the future. This prevents us from having to
+             * iterate over the whole queue on every refresh at the cost of some requests having to wait an extra cycle if they get stuck
+             * behind a request that missed the refresh cycle.
+             */
+            Iterator<Tuple<Translog.Location, Consumer<Boolean>>> itr = refreshListeners.iterator();
+            while (itr.hasNext()) {
+                Tuple<Translog.Location, Consumer<Boolean>> listener = itr.next();
+                if (listener.v1().compareTo(currentRefreshLocation) > 0) {
```

Comment:
This design came from a talk with @jasontedor. He wasn't a fan of the synchronized block that using a non-concurrent data structure required.

Originally I was just going over the whole list and only calling the listeners that were made visible - presumably that is most of them most of the time. I was using a LinkedList because I was removing as I went but copy-to-ArrayList is probably better.

This particular tradeoff is one he seemed to really like: no locks, inaccurate counts, and listeners potentially having to wait for the next refresh if they get out of order. I'll defer to smarter people what the right tradeoffs are, I just wanted to make it clear why I made the ones that I did.